tensorflow/compiler/jit/deadness_analysis.cc

```diff
@@ -914,10 +954,135 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+          "BuildControlFlowInfo");
+    }
+  }
+
+  *frame = cfi_iter->frame_name;
+  return Status::OK();
+}
+
+// Compute a special topological order for the Graph, where nodes having the
+// same root frame are placed adjacent to each other.  The traversal uses a
+// variant of Kahn's algorithm.  num_ready_inputs is used to keep track of how
+// many inputs of each node are ready; a node is ready to be scheduled if all
+// of its inputs are ready.
+// Ref. to https://en.wikipedia.org/wiki/Topological_sorting for details.
+Status GetFrameBasedTopologicalOrder(const Graph* g,
+                                     absl::Span<const ControlFlowInfo> cf_infos,
+                                     std::vector<Node*>* order) {
+  absl::flat_hash_map<absl::string_view, size_t> num_enters;
+  absl::flat_hash_map<absl::string_view, size_t> num_exits;
+  std::vector<size_t> num_ready_inputs(g->num_node_ids(), 0);
+  Node* src_node = g->source_node();
+  for (const auto* node : g->op_nodes()) {
+    const ControlFlowInfo& cf = cf_infos[node->id()];
+    bool is_root_level = cf.parent_frame == src_node;
+    if (IsEnter(node) && is_root_level) {
+      // Since we care only the root-level frame, full frame names are the same
+      // as frame names.
+      ++num_enters[cf.frame_name];
+    } else if (IsExit(node) && is_root_level) {
+      ++num_exits[cf.frame_name];
+    }
+    // Edge NextIteration->Merge is counted before starting the traveral to
+    // break the backedges.
+    if (IsMerge(node)) {
+      for (const Edge* e : node->in_edges()) {
+        if (IsNextIteration(e->src())) {
+          ++num_ready_inputs[node->id()];
+        }
+      }
+    }
+  }
+
+  std::deque<Node*> ready;
+  ready.push_back(src_node);
+  // ready_enters_per_frame and ready_exits serve as a staging area to buffer
+  // the ready enters/exits before they are moved to the `ready` queue for
+  // controlling the start and end of a processing frame.
+  absl::flat_hash_map<string, std::vector<Node*>> ready_enters_per_frame;
+  // Exit nodes shall all be from the same frame, as we process a frame at a
+  // time. So, one vector is enough.
+  std::vector<Node*> ready_exits;
+  while (!ready.empty()) {
+    Node* curr_node = ready.front();
+    ready.pop_front();
+
+    VLOG(4) << "Visiting " << curr_node->name();
+    order->push_back(curr_node);
+
+    for (const Edge* out_edge : curr_node->out_edges()) {
+      Node* out = out_edge->dst();
+      int out_id = out->id();
+      if (IsNextIteration(curr_node) && IsMerge(out)) {
+        // Edge NextIteration->Merge has been counted.
+        continue;
+      }
+      ++num_ready_inputs[out->id()];
+      if (!out->IsOp()) continue;  // Skip Sink/Source nodes.
+      if (num_ready_inputs[out->id()] != out->in_edges().size()) continue;
+
+      bool is_root_level = cf_infos[out_id].parent_frame == src_node;
+      string frame_name = cf_infos[out_id].frame_name;
+      if (IsEnter(out) && is_root_level) {
```

Comment:
How about extracting out `IsEnter(out) && is_root_level` into a `IsRootEnter`?  Similarly for `IsRootExit`?

Reply:
I don't see a good way to do that. If we make `IsRootExit()` a function, we need to pass `control_flow_info_` to it. Wouldn't that be a bit cumbersome?

Reply:
These would have to be member functions on `DeadnessAnalysisImpl` to avoid having to pass around `control_flow_info_` (which I agree would be cumbersome).